File: app/iOS/Thermal Comfort Study/AdaptiveComfortActionItem.swift
//
//  AdaptiveComfortActionItem.swift
//  Thermal Comfort Study
//
//  Created by Madhav Achar on 10/19/16.
//  Copyright © 2016 boschsc. All rights reserved.
//

import Foundation

class AdaptiveComfortActionItem: NSObject {
    
    private var username: String?
    private var timestamp: String?
    private var action: String?
    private var actionDescription: String?
    
    public static let ELEMENT = "item"
    public static let ITEM_ID_PREFIX = "_ActionData"
    
    // Expects nil for description if action is not "Other"
    init(username: String, timestamp: Date, action: String, description: String?) {
        self.username = username
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSSZ"
        self.timestamp = dateFormatter.string(from: timestamp)
        
        self.action = action
        self.actionDescription = description
    }
    
    public func toXML() -> String {
        // If action equals other, set value to description
        var value = action!
        if let description = actionDescription {
            value = description
        }
        
        var str=""
        str.append("<item id=\"" + getId() + "\">")
        str.append("<transducerData id=\"occthermcomf_bosch\" ")
        str.append("type=\"TCS\" ")
        str.append("class=\"ActionData\" ")
        str.append("username=\"\(username!)\" ")
        str.append("timestamp=\"\(timestamp!)\" ")
        str.append("name=\"AdaptiveComfortAction\" ")
        str.append("value=\"\(value)\"/>")
        str.append("</" + AdaptiveComfortActionItem.ELEMENT + ">")
    
        return str
    }
    
    public func getId() -> String {
        return (AdaptiveComfortActionItem.ITEM_ID_PREFIX + timestamp!)
    }
}

File: app/android/app/src/main/java/bosch/smartcampus/thermalcomfortstudy/packet/AdaptiveComfortActionItem.java
package bosch.smartcampus.thermalcomfortstudy.packet;

import org.jivesoftware.smackx.pubsub.Item;

import bosch.smartcampus.thermalcomfortstudy.lib.AdaptiveComfortAction;
import bosch.smartcampus.thermalcomfortstudy.lib.Timestamp;

/**
 * Created by rsukkerd on 7/5/16.
 */
public class AdaptiveComfortActionItem extends Item {
    public static final String NAMESPACE = null;
    public static final String ELEMENT = "item";
    public static final String ITEM_ID_PREFIX = "_ActionData";

    public static final String TRANSDUCER_DATA_ID = "occthermcomf_bosch";
    public static final String TRANSDUCER_DATA_TYPE = "TCS";
    public static final String TRANSDUCER_DATA_CLASS = "ActionData";

    private static final String ADAPTIVE_COMFORT_ACTION_NAME = "AdaptiveComfortAction";

    private String username;
    private Timestamp timestamp;
    private AdaptiveComfortAction action;
    private String actionDescription;

    public AdaptiveComfortActionItem(String username, Timestamp timestamp,
                                     AdaptiveComfortAction action) {
        this(username, timestamp, action, "");
    }

    public AdaptiveComfortActionItem(String username, Timestamp timestamp,
                                     AdaptiveComfortAction action, String actionDescription) {
        this.username = username;
        this.timestamp = timestamp;
        this.action = action;
        this.actionDescription = actionDescription;
    }

    public String getUsername() {
        return username;
    }

    public Timestamp getTimestamp() {
        return timestamp;
    }

    public AdaptiveComfortAction getAdaptiveComfortAction() {
        return action;
    }

    public String getActionDescription() {
        return actionDescription;
    }

    @Override
    public String getNamespace() {
        return NAMESPACE;
    }

    @Override
    public String getElementName() {
        return ELEMENT;
    }

    @Override
    public String getId() {
        return ITEM_ID_PREFIX + timestamp;
    }

    @Override
    public String toXML() {
        StringBuilder builder = new StringBuilder();
        builder.append("<" + ELEMENT + " id=\"" + getId() + "\">");

        if (action == AdaptiveComfortAction.OTHER) {
            builder.append(getTransducerDataElement(ADAPTIVE_COMFORT_ACTION_NAME, actionDescription));
        } else {
            builder.append(getTransducerDataElement(ADAPTIVE_COMFORT_ACTION_NAME, action.toString()));
        }

        builder.append("</" + ELEMENT + ">");
        return builder.toString();
    }

    protected String getTransducerDataElement(String name, String value) {
        String element = String.format("<transducerData id=\"%s\" type=\"%s\" class=\"%s\" ",
                TRANSDUCER_DATA_ID, TRANSDUCER_DATA_TYPE, TRANSDUCER_DATA_CLASS);
        element += "username=\"" + username + "\" ";
        element += "timestamp=\"" + timestamp + "\" ";
        element += "name=\"" + name + "\" ";
        element += "value=\"" + value + "\"/>";
        return element;
    }
}
